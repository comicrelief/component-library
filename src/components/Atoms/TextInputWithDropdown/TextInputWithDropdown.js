import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Text from '../Text/Text';
import {
  Container,
  Dropdown,
  DropdownList,
  DropdownItem,
  DropdownItemSelectable,
  TextItalic,
  InputWithSpaceAfterLabel
} from './TextInputWithDropdown.style';

const KEY_CODE_ENTER = 13;
const KEY_CODE_SPACE = 32;
const KEY_CODE_UP = 38;
const KEY_CODE_DOWN = 40;
const KEY_CODE_ESCAPE = 27;

/**
 * This component deals with the visual aspect of a text input with typeahead-style functionality
 *
 * Proper functionality / behaviour (e.g. API querying, state management, etc.) are handled by the parent
 *  component (via the value, options, onChange and onSelect props)
 *
 * See the Typeahead and SchoolLookup molecules for the full implementation
 */
const TextInputWithDropdown = React.forwardRef(
  (
    {
      value,
      options,
      onChange,
      onSelect,
      id,
      name,
      label,
      dropdownInstruction,
      className,
      ...otherInputProps
    },
    ref
  ) => {
    const [activeOption, setActiveOption] = useState(-1);
    const [forceClosed, setForceClosed] = useState(false);
    useEffect(() => {
      // reset if options change
      setActiveOption(-1);
      setForceClosed(false);
    }, [options]);

    const down = () =>
      activeOption < options.length - 1
        ? setActiveOption(activeOption + 1)
        : setActiveOption(0);
    const up = () =>
      activeOption < 1
        ? setActiveOption(options.length - 1)
        : setActiveOption(activeOption - 1);

    const navigateOptions = e => {
      if (options.length === 0) {
        return;
      }

      const keyCode = e.keyCode || e.which;

      if (keyCode === KEY_CODE_DOWN) {
        e.preventDefault();
        down();
      } else if (keyCode === KEY_CODE_UP) {
        e.preventDefault();
        up();
      } else if (keyCode === KEY_CODE_ESCAPE) {
        setForceClosed(true);
      }
    };

    const inputProps = {
      value,
      onChange,
      id,
      name,
      label,
      autoComplete: 'off',
      type: 'text',
      ...otherInputProps
    };

    const optionsProps = {
      options,
      onSelect,
      dropdownInstruction,
      activeOption,
      resetActiveOption: () => setActiveOption(-1)
    };

    return (
      <Container
        className={`TextInputWithDropdown ${className}`.trim()}
        onKeyDown={navigateOptions}
      >
        <InputWithSpaceAfterLabel
          {...inputProps}
          className="TextInputWithDropdown__input"
          ref={ref}
        />
        {options.length > 0 && forceClosed === false && (
          <Options
            {...optionsProps}
            className="TextInputWithDropdown__options"
          />
        )}
      </Container>
    );
  }
);

const Options = ({
  options,
  dropdownInstruction,
  onSelect,
  activeOption,
  resetActiveOption,
  ...rest
}) => {
  // Reset 'activeOption' when the list is unfocused.
  const onBlur = e => {
    const { target } = e;
    // There's a delay before the new activeOption becomes the document.activeElement when
    //  scrolling through the dropdown list via keyboard.
    setTimeout(() => {
      if (document.activeElement.parentNode !== target.parentNode) {
        resetActiveOption();
      }
    }, 100);
  };

  return (
    <Dropdown {...rest} tabIndex="0">
      <DropdownList
        role="listbox"
        onBlur={onBlur}
        aria-activedescendant={
          activeOption > -1 ? `option-${activeOption}` : undefined
        }
      >
        {dropdownInstruction && (
          <DropdownItem role="option">
            <TextItalic>{dropdownInstruction}</TextItalic>
          </DropdownItem>
        )}
        {options.map((option, index) => (
          <DropdownItemSelectable
            id={`option-${index}`}
            role="option"
            key={option}
            onClick={() => onSelect(option, index)}
            onKeyPress={e => {
              const keyCode = e.keyCode || e.which;
              if (keyCode === KEY_CODE_SPACE || keyCode === KEY_CODE_ENTER) {
                onSelect(option, index);
              }
            }}
            tabIndex="-1"
            aria-selected={index === activeOption}
            ref={
              index === activeOption
                ? element => element && element.focus()
                : null
            }
          >
            <Text>{option}</Text>
          </DropdownItemSelectable>
        ))}
      </DropdownList>
    </Dropdown>
  );
};

TextInputWithDropdown.propTypes = {
  value: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  className: PropTypes.string,
  dropdownInstruction: PropTypes.string
};

TextInputWithDropdown.defaultProps = {
  dropdownInstruction: null,
  className: ''
};

Options.propTypes = {
  options: PropTypes.arrayOf(PropTypes.string).isRequired,
  onSelect: PropTypes.func.isRequired,
  dropdownInstruction: PropTypes.string,
  activeOption: PropTypes.number.isRequired,
  resetActiveOption: PropTypes.func.isRequired
};

Options.defaultProps = {
  dropdownInstruction: null
};

export default TextInputWithDropdown;
