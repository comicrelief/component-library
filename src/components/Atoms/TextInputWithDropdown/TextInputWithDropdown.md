# TextInputWithDropdown

```jsx
const [value, setValue] = React.useState('');
const [options, setOptions] = React.useState([]);

const dummyOptions = [
  'Option 1',
  'Another option',
  'And another',
  'Click me',
  'No, click me',
];

<TextInputWithDropdown
  value={value}
  options={options}
  onChange={(e) => {
    setValue(e.currentTarget.value);
    setOptions(value.trim().length > 2 ? dummyOptions : []);
  }}
  onSelect={(value, index) => {
    alert(`Selected [${value}] at index [${index}]`);
    setValue('');
    setOptions([]);
  }}
  id="input-with-dropdown"
  name="query"
  label="Text input with dropdown"
  dropdownInstruction="Choose one from the list below..."
/>;
```
