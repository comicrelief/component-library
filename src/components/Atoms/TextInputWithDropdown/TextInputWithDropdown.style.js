import styled from 'styled-components';

import spacing from '../../../theme/shared/spacing';
import zIndex from '../../../theme/shared/zIndex';
import Text from '../Text/Text';

const Container = styled.div`
  position: relative;
`;

const Dropdown = styled.div`
  ${zIndex('high')}
  font-family: ${({ theme }) => theme.fontFamilies('Montserrat')};
  position: absolute;
  left: 0;
  max-height: 300px;
  overflow: auto;
  background-color: ${({ theme }) => theme.color('white')};
  border: 1px solid;
  margin-top: -1px;
  width: 100%;
  @media ${({ theme }) => theme.allBreakpoints('M')} {
    max-width: 500px;
  }
`;

const DropdownList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
`;

const DropdownItem = styled.li`
  padding: ${spacing('sm')};
`;

const DropdownItemSelectable = styled(DropdownItem)`
  cursor: pointer;
  border-top: 1px solid ${({ theme }) => theme.color('grey_light')};

  &:hover,
  &:focus {
    background-color: ${({ theme }) => theme.color('grey_light')};
  }
`;

const TextItalic = styled(Text)`
  font-style: italic;
`;

export {
  Container,
  Dropdown,
  DropdownList,
  DropdownItem,
  DropdownItemSelectable,
  TextItalic
};
