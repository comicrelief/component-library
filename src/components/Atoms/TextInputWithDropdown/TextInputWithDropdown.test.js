import React from 'react';
import 'jest-styled-components';

import renderWithTheme from '../../../hoc/shallowWithTheme';
import TextInputWithDropdown from './TextInputWithDropdown';
import { Dropdown, Container } from './TextInputWithDropdown.style';

it('renders correctly with no value and no options', () => {
  const renderer = renderWithTheme(
    <TextInputWithDropdown
      value=""
      id="text-input-with-dropdown"
      label="Search for bikes"
      name="query"
      onChange={() => {}}
      onSelect={() => {}}
      options={[]}
    />
  );

  expect(renderer.toJSON()).toMatchSnapshot();
  expect(() => renderer.root.findByType(Container).findByType(Dropdown)).toThrow();
});

it('renders correctly with value and no options', () => {
  const renderer = renderWithTheme(
    <TextInputWithDropdown
      value="bikes"
      id="text-input-with-dropdown"
      label="Search for bikes"
      name="query"
      onChange={() => {}}
      onSelect={() => {}}
      options={[]}
    />
  );

  expect(renderer.toJSON()).toMatchSnapshot();
  expect(() => renderer.root.findByType(Container).findByType(Dropdown)).toThrow();
});

it('renders correctly with value and options', () => {
  const renderer = renderWithTheme(
    <TextInputWithDropdown
      value="c"
      id="text-input-with-dropdown"
      label="Search for bikes"
      name="query"
      onChange={() => {}}
      onSelect={() => {}}
      options={['Canyon', 'Cannondale', 'Condor', 'Cube']}
    />
  );

  expect(renderer.toJSON()).toMatchSnapshot();
  expect(renderer.root.findByType(Container).findByType(Dropdown)).toBeTruthy();
});
