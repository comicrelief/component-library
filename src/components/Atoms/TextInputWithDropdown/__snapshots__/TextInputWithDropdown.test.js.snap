// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no value and no options 1`] = `
.c2 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  font-weight: bold;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #5C5C5E;
  width: 100%;
}

.c3 {
  margin-bottom: 0.5rem;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 48px;
  padding: 1rem 1.5rem;
  background-color: #F4F3F5;
  border: 1px solid;
  border-color: #E1E2E3;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #000000;
  border-radius: 0.5rem;
  font-size: inherit;
  z-index: 2;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c5:focus {
  border: 1px solid #666;
}

.c4 {
  position: relative;
  font-size: 1.25rem;
}

.c0 {
  position: relative;
}

@media (min-width:740px) {
  .c5 {
    max-width: 290px;
  }
}

<div
  className="c0 TextInputWithDropdown"
  onKeyDown={[Function]}
>
  <label
    className="c1 TextInputWithDropdown__input"
    htmlFor="text-input-with-dropdown"
  >
    <span
      className="c2 c3"
      color="inherit"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Search for bikes",
        }
      }
      size="s"
    />
    <div
      className="c4"
    >
      
      <input
        aria-describedby="text-input-with-dropdown"
        autoComplete="off"
        className="c5"
        id="text-input-with-dropdown"
        name="query"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
        value=""
      />
    </div>
    
  </label>
</div>
`;

exports[`renders correctly with value and no options 1`] = `
.c2 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  font-weight: bold;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #5C5C5E;
  width: 100%;
}

.c3 {
  margin-bottom: 0.5rem;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 48px;
  padding: 1rem 1.5rem;
  background-color: #F4F3F5;
  border: 1px solid;
  border-color: #E1E2E3;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #000000;
  border-radius: 0.5rem;
  font-size: inherit;
  z-index: 2;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c5:focus {
  border: 1px solid #666;
}

.c4 {
  position: relative;
  font-size: 1.25rem;
}

.c0 {
  position: relative;
}

@media (min-width:740px) {
  .c5 {
    max-width: 290px;
  }
}

<div
  className="c0 TextInputWithDropdown"
  onKeyDown={[Function]}
>
  <label
    className="c1 TextInputWithDropdown__input"
    htmlFor="text-input-with-dropdown"
  >
    <span
      className="c2 c3"
      color="inherit"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Search for bikes",
        }
      }
      size="s"
    />
    <div
      className="c4"
    >
      
      <input
        aria-describedby="text-input-with-dropdown"
        autoComplete="off"
        className="c5"
        id="text-input-with-dropdown"
        name="query"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
        value="bikes"
      />
    </div>
    
  </label>
</div>
`;

exports[`renders correctly with value and options 1`] = `
.c2 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  font-weight: bold;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c10 {
  font-size: 1rem;
  line-height: 1rem;
  text-transform: inherit;
  line-height: normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: #5C5C5E;
  width: 100%;
}

.c3 {
  margin-bottom: 0.5rem;
}

.c5 {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  height: 48px;
  padding: 1rem 1.5rem;
  background-color: #F4F3F5;
  border: 1px solid;
  border-color: #E1E2E3;
  box-shadow: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #000000;
  border-radius: 0.5rem;
  font-size: inherit;
  z-index: 2;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

.c5:focus {
  border: 1px solid #666;
}

.c4 {
  position: relative;
  font-size: 1.25rem;
}

.c0 {
  position: relative;
}

.c6 {
  z-index: 3;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
  position: absolute;
  left: 0;
  max-height: 300px;
  overflow: auto;
  background-color: #FFFFFF;
  border: 1px solid;
  margin-top: -1px;
  width: 100%;
}

.c7 {
  list-style: none;
  padding: 0;
  margin: 0;
}

.c8 {
  padding: 0.5rem;
}

.c9 {
  cursor: pointer;
  border-top: 1px solid #F4F3F5;
}

.c9:hover,
.c9:focus {
  background-color: #F4F3F5;
}

@media (min-width:740px) {
  .c5 {
    max-width: 290px;
  }
}

@media (min-width:740) {
  .c6 {
    max-width: 500px;
  }
}

<div
  className="c0 TextInputWithDropdown"
  onKeyDown={[Function]}
>
  <label
    className="c1 TextInputWithDropdown__input"
    htmlFor="text-input-with-dropdown"
  >
    <span
      className="c2 c3"
      color="inherit"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Search for bikes",
        }
      }
      size="s"
    />
    <div
      className="c4"
    >
      
      <input
        aria-describedby="text-input-with-dropdown"
        autoComplete="off"
        className="c5"
        id="text-input-with-dropdown"
        name="query"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
        value="c"
      />
    </div>
    
  </label>
  <div
    className="c6 TextInputWithDropdown__options"
  >
    <ul
      className="c7"
      onBlur={[Function]}
      role="listbox"
    >
      <li
        aria-selected={false}
        className="c8 c9"
        id="option-0"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="option"
        tabIndex="-1"
      >
        <span
          className="c10"
          color="inherit"
          size="s"
        >
          Canyon
        </span>
      </li>
      <li
        aria-selected={false}
        className="c8 c9"
        id="option-1"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="option"
        tabIndex="-1"
      >
        <span
          className="c10"
          color="inherit"
          size="s"
        >
          Cannondale
        </span>
      </li>
      <li
        aria-selected={false}
        className="c8 c9"
        id="option-2"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="option"
        tabIndex="-1"
      >
        <span
          className="c10"
          color="inherit"
          size="s"
        >
          Condor
        </span>
      </li>
      <li
        aria-selected={false}
        className="c8 c9"
        id="option-3"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="option"
        tabIndex="-1"
      >
        <span
          className="c10"
          color="inherit"
          size="s"
        >
          Cube
        </span>
      </li>
    </ul>
  </div>
</div>
`;
